Signal/src/util/OWSBackupExportJob.m

```diff
@@ -523,47 +701,69 @@ - (BOOL)saveNextAttachmentFileToCloud:(OWSBackupJobCompletion)completion
                 }
-                strongSelf.attachmentRecordMap[recordName] = attachmentExport.relativeFilePath;
-                DDLogVerbose(@"%@ exported attachment: %@ as %@",
+
+                if (![attachmentExport cleanUp]) {
+                    DDLogError(@"%@ couldn't clean up attachment export.", self.logTag);
+                    // Attachment files are non-critical so any error uploading them is recoverable.
+                }
+
+                OWSBackupExportItem *exportItem = [OWSBackupExportItem new];
+                exportItem.encryptedItem = attachmentExport.encryptedItem;
+                exportItem.recordName = recordName;
+                exportItem.attachmentExport = attachmentExport;
+                [strongSelf.savedAttachmentItems addObject:exportItem];
+
+                DDLogVerbose(@"%@ saved attachment: %@ as %@",
                     self.logTag,
-                    attachmentFilePath,
+                    attachmentExport.attachmentFilePath,
                     attachmentExport.relativeFilePath);
-                [strongSelf saveNextFileToCloud:completion];
+                [strongSelf saveNextFileToCloudWithCompletion:completion];
             });
         }
         failure:^(NSError *error) {
             // Ensure that we continue to work off the main thread.
             dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                if (![attachmentExport cleanUp]) {
+                    DDLogError(@"%@ couldn't clean up attachment export.", self.logTag);
+                    // Attachment files are non-critical so any error uploading them is recoverable.
+                }
+
                 // Attachment files are non-critical so any error uploading them is recoverable.
-                [weakSelf saveNextFileToCloud:completion];
+                [weakSelf saveNextFileToCloudWithCompletion:completion];
             });
         }];
+
     return YES;
 }
 
-- (void)saveManifestFileToCloud:(OWSBackupJobCompletion)completion
+- (void)saveManifestFileToCloudWithCompletion:(OWSBackupJobCompletion)completion
 {
     OWSAssert(completion);
 
-    if (![self writeManifestFile]) {
+    OWSBackupEncryptedItem *_Nullable encryptedItem = [self writeManifestFile];
+    if (!encryptedItem) {
         completion(OWSErrorWithCodeDescription(OWSErrorCodeExportBackupFailed,
             NSLocalizedString(@"BACKUP_EXPORT_ERROR_COULD_NOT_EXPORT",
                 @"Error indicating the a backup export could not export the user's data.")));
         return;
     }
-    OWSAssert(self.manifestFilePath);
+
+    OWSBackupExportItem *exportItem = [OWSBackupExportItem new];
+    exportItem.encryptedItem = encryptedItem;
 
     __weak OWSBackupExportJob *weakSelf = self;
 
-    [OWSBackupAPI upsertManifestFileToCloudWithFileUrl:[NSURL fileURLWithPath:self.manifestFilePath]
+    [OWSBackupAPI upsertManifestFileToCloudWithFileUrl:[NSURL fileURLWithPath:encryptedItem.filePath]
```

Comment:
just geeking out on API, but if the OWSBackupAPI upsert/upload api took an EncryptedItem instead of a filePath, we could use the compiler to help make sure were doing everything in the right order (encrypting before uploading). TIOLI of course.